### PR TITLE
Addressing Nell's feedback

### PR DESCRIPTION
**Pull request by Artyom17**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -41,22 +29,32 @@ One of the very common use cases of WebXR is 180 and/or 360 photos or videos, bo
 Another example where composition layers are going to shine is displaying text or high resolution textures in the virtual world. Since composition layers allow to sample source texture at full resolution without downsampling to the resolution of the eye buffers (which is usually much lower than the physical resolution of the device's screen(s)) the readability of the text is going to be significantly improved.
 
 
-### Implementation overview
+## Goals of the proposal
+The purpose of this document is to unlock the ability for User Agents to experiment with multiple layers. 
+* Introduce a concept of layers into WebXR spec and the way how layers should be provided from user's code to Browser;
```
Comment (NellWaliczek) [inline README.md]: The concept of layers already exists in the WebXR spec.  Did you mean the concept of multiple layers?

File: README.md [hunk 2]
```diff
@@ -41,22 +29,32 @@ One of the very common use cases of WebXR is 180 and/or 360 photos or videos, bo
...
 viewer.
 * Introduce a way to determine which layers are supported and what is the maximum amount of the layers supported (fingerprinting?);
 * Add common properties to [XRLayer](https://immersive-web.github.io/webxr/#xrlayer-interface) (such as chromatic aberration or alpha blending flags);
 * Introduce a concept of "image source" for layers and a way to create / request image sources from the UA;
-* Optionally, change [XRWebGLLayer](https://immersive-web.github.io/webxr/#xrwebgllayer-interface) to comply with the new concepts of XRLayer and image source. Or, we can leave it as is and replace it with `XRLayerProjection` once we are ready;
-* Introduce different subtypes to [XRLayer](https://immersive-web.github.io/webxr/#xrlayer-interface) which will have all the necessary attributes for each layer.
+* Optionally, change [XRWebGLLayer](https://immersive-web.github.io/webxr/#xrwebgllayer-interface) to comply with the new concepts of XRLayer and image source. Or, we can leave it as is and replace it with a different layer once we are ready.
```
Comment (NellWaliczek) [inline README.md]: For now, let's leave off these last two bullets entirely.  We can mention them verbally in the session tomorrow, but I think they're unfinished enough that it'll be distracting here.  Once they're in better shape, we can add the reference back if we think it's valuable to continue editing this document

File: README.md [hunk 3]
```diff
@@ -41,22 +29,32 @@ One of the very common use cases of WebXR is 180 and/or 360 photos or videos, bo
...
 
-All of these changes could be split into two categories: "breaking changes" and "additive changes".
-
-### Breaking changes
+The proposed IDL at the end of this document is for a reference only and not final by any means.
 
 #### Changes in XRRenderState
```
Comment (NellWaliczek) [inline README.md]: Should this be an H3?  Also change the others?

File: README.md [hunk 4]
```diff
@@ -74,210 +72,41 @@ dictionary XRRenderStateInit {
 };
 ```
 
-##### Drawing layers 
-As was mentioned before, the drawing algorithm is as simple as a "painter's algorithm" where layers render in order of their appearance in the sequence, with the 0th layer drawn first (usually, the 0th layer is the eye-buffer one). Each successive layer potentially overwriting the destination layers whether or not the new layers are virtually closer to the viewer. There is no depth testing between the layers. 
+#### Drawing layers
```
Comment (NellWaliczek) [inline README.md]: Maybe call this "Layer composition"

File: README.md [hunk 5]
```diff
@@ -74,210 +72,41 @@ dictionary XRRenderStateInit {
...
-* `blendTextureSourceAlpha` - enables the layer's texture alpha channel; when it is set to `true` then the layer composition uses the alpha channel for the blending of the layer's image against the destination. The image's color channels must be encoded with premultiplied alpha. When the image has no alpha channel then this flag is ignored. If this flag is not set then the image is treated as if each texel is opaque, regardless of the presence of an alpha channel.
-The blending opration between the source and destination is an addition. The formula for the blending of each color component is as follows: `destination_color = (source_color + (destination_color * (1 - source_alpha)))`.
-> **TODO** Verify with WebGL folks
-
-##### Added methods to XRLayer
-* `void requestUpdate()` - the method that should be called to indicate the layer's content changed; by default, browser should assume that the content of the layers (image source) is not updated.
-> **TODO** Should we add a way to control layer's visibility or, just use `XRSession/updateRenderState` if layer's visibility changes?
+#### Additions to XRLayer / XRLayerInit
```
Comment (NellWaliczek) [inline README.md]: I'd remove this section entirely, tbh :)

File: README.md [hunk 6]
```diff
@@ -74,210 +72,41 @@ dictionary XRRenderStateInit {
...
+#### Drawing layers 
+As was mentioned before, the drawing algorithm is a simple "painter's algorithm" where layers render in order of their appearance in the sequence, with the 0th layer drawn first (usually, the 0th layer is the eye-buffer one). Each successive layer potentially overwriting the destination layers whether or not the new layers are virtually closer to the viewer. There is no depth testing between the layers. 
 Layer's visibility is controlled by presence (or absence) the layer in the sequence. To hide or show a layer - call [updateRenderState](https://immersive-web.github.io/webxr/#dom-xrsession-updaterenderstate) and provide a new sequence of layers.
 
-#### Optional changes to XRWebGLLayer
-Having XRLayerSourceImage concept introduced, shouldn't we modify XRWebGLLayer to use it instead of explicit reference to framebuffer or texture array (for the XRWebGLArrayLayer)? By doing this, we could avoid introducing an extra XRWebGLArrayLayer type for multiview support in this case. Alternatively, we can introduce a new layer type - `XRLayerProjection`, similarly to what OpenXR has.
+#### Adding layers capabilities
```
Comment (NellWaliczek) [inline README.md]: Maybe rename to "Querying layer capabilities"

File: README.md [hunk 7]
```diff
@@ -74,210 +72,41 @@ dictionary XRRenderStateInit {
...
-The blending opration between the source and destination is an addition. The formula for the blending of each color component is as follows: `destination_color = (source_color + (destination_color * (1 - source_alpha)))`.
-> **TODO** Verify with WebGL folks
-
-##### Added methods to XRLayer
-* `void requestUpdate()` - the method that should be called to indicate the layer's content changed; by default, browser should assume that the content of the layers (image source) is not updated.
-> **TODO** Should we add a way to control layer's visibility or, just use `XRSession/updateRenderState` if layer's visibility changes?
+#### Additions to XRLayer / XRLayerInit
+Some common properties and methods could be added to XRLayer / XRLayerInit. See [here](details.md) for more details.
 
 #### Layer image source
```
Comment (NellWaliczek) [inline README.md]: Also would remove this entirely :)

File: README.md [hunk 8]
```diff
@@ -74,210 +72,41 @@ dictionary XRRenderStateInit {
...
-```webidl
-[ SecureContext, Exposed=Window ] interface XRLayerSubImage {
-  readonly attribute XRLayerImageSource imageSource;
-  readonly attribute FrozenArray<float> imageRectUV; // 2D rect in UV space
-  readonly attribute long imageArrayIndex;
-};
-```
-### Proposed types of layers
-Not all layers are going to be supported by all hardware/browsers. We would need to figure out the bare minimum of layer types to be supported. I have the following ones in mind: the transparent or opaque quadrilateral, cubemap, cylindrical and equirect layers. Additionally, we might want to replace the `XRWebGLLayer` by the `XRLayerProjection` one.
+#### Optional changes to XRWebGLLayer
```
Comment (NellWaliczek) [inline README.md]: And this! :)

File: README.md [hunk 9]
```diff
@@ -74,210 +72,41 @@ dictionary XRRenderStateInit {
...
-[ SecureContext, Exposed=Window ] interface XRWebGLLayer : XRLayer {
-  readonly attribute XRWebGLRenderingContext context;
-  readonly attribute XRLayerSourceImage imageSource;
-
-  XRViewport? getViewport(XRView view);
-
-  static double getNativeFramebufferScaleFactor(XRSession session);
-};
 ```
+More complete IDL (but still - very much preliminary and WIP) could be found [here](details.md).
```
Comment (NellWaliczek) [inline README.md]:
```suggestion
Based on the foundation described in this document, a next set of functionality is being explored [here](details.md) that goes into more details about common parts of a multilayer system and enumerates potential layer types.```

Comment (NellWaliczek) [inline README.md]: Maybe just move this to the todo section?

File: details.md [hunk 10]
```diff
@@ -0,0 +1,196 @@
+### Proposed types of layers
```
Comment (NellWaliczek) [inline details.md]: I'm super looking forward to working on this design when we get closure on the next steps for multilayer!

File: README.md (hunk 1 above)
Comment (Artyom17) [reply README.md]: yes

**Comment by Artyom17** [comment]
Committing the PR  + additional changes suggested by Nell for the first commit

**Pull request merged**